File: modules/common-model/src/main/java/com/shterneregen/securelan/common/model/FileTransferProgress.java
package com.shterneregen.securelan.common.model;

import java.util.Objects;

public record FileTransferProgress(
        String transferId,
        long transferredBytes,
        long totalBytes,
        TransferStatus status
) {
    public FileTransferProgress {
        Objects.requireNonNull(transferId, "transferId must not be null");
        Objects.requireNonNull(status, "status must not be null");
        if (transferredBytes < 0) {
            throw new IllegalArgumentException("transferredBytes must not be negative");
        }
        if (totalBytes < 0) {
            throw new IllegalArgumentException("totalBytes must not be negative");
        }
        if (totalBytes > 0 && transferredBytes > totalBytes) {
            throw new IllegalArgumentException("transferredBytes must not exceed totalBytes");
        }
    }

    public int percent() {
        if (totalBytes <= 0) {
            return 0;
        }
        return (int) ((transferredBytes * 100) / totalBytes);
    }
}

File: modules/common-model/src/main/kotlin/com/shterneregen/securelan/common/model/FileTransferProgress.kt
package com.shterneregen.securelan.common.model

import java.util.Objects

@JvmRecord
data class FileTransferProgress(
    val transferId: String,
    val transferredBytes: Long,
    val totalBytes: Long,
    val status: TransferStatus,
) {
    init {
        Objects.requireNonNull(transferId, "transferId must not be null")
        Objects.requireNonNull(status, "status must not be null")
        require(transferredBytes >= 0) { "transferredBytes must not be negative" }
        require(totalBytes >= 0) { "totalBytes must not be negative" }
        require(!(totalBytes > 0 && transferredBytes > totalBytes)) { "transferredBytes must not exceed totalBytes" }
    }

    fun percent(): Int = if (totalBytes <= 0) 0 else ((transferredBytes * 100) / totalBytes).toInt()
}
